File: frontend/src/lib/placeholderLabels.js
const ORDINALS = ["", "1st", "2nd", "3rd", "4th"];

function knockoutOutcomeLabel(matchNum, outcome) {
  const role = outcome === "winner" ? "Winner" : "Loser";
  if (matchNum >= 101 && matchNum <= 102) return `${role} of semi-final ${matchNum - 100}`;
  if (matchNum >= 97 && matchNum <= 100) return `${role} of quarter-final ${matchNum - 96}`;
  if (matchNum >= 89 && matchNum <= 96) return `${role} of Round of 16 (match ${matchNum})`;
  if (matchNum >= 73 && matchNum <= 88) return `${role} of Round of 32 (match ${matchNum})`;
  return `${role} of match ${matchNum}`;
}

export function placeholderLabel(code) {
  if (!code) return "To be decided";

  const text = String(code).trim();
  if (!text) return "To be decided";

  const groupPos = text.match(/^([1-4])([A-L])$/);
  if (groupPos) {
    return `${ORDINALS[Number(groupPos[1])]} in Group ${groupPos[2]}`;
  }

  const bestThird = text.match(/^3([A-L]+)$/);
  if (bestThird) {
    const groups = [...bestThird[1]].join(", ");
    return `Best 3rd-place team (groups ${groups})`;
  }

  const altGroup = text.match(/^([A-L])([1-4])$/);
  if (altGroup) {
    return `${ORDINALS[Number(altGroup[2])]} in Group ${altGroup[1]}`;
  }

  const winner = text.match(/^W(\d+)$/);
  if (winner) return knockoutOutcomeLabel(Number(winner[1]), "winner");

  const loser = text.match(/^RU(\d+)$/) || text.match(/^L(\d+)$/);
  if (loser) return knockoutOutcomeLabel(Number(loser[1]), "loser");

  return text;
}

export function teamDisplayName(team) {
  if (!team) return "To be decided";
  if (team.status === "placeholder") return placeholderLabel(team.draw_name || team.display_name);
  return team.display_name || team.draw_name || "To be decided";
}

File: scripts/placeholder_labels.py
"""Turn FIFA bracket placeholder codes into plain English."""

from __future__ import annotations

import re

ORDINALS = ("", "1st", "2nd", "3rd", "4th")

GROUP_POSITION_RE = re.compile(r"^([1-4])([A-L])$")
BEST_THIRD_RE = re.compile(r"^3([A-L]+)$")
ALT_GROUP_RE = re.compile(r"^([A-L])([1-4])$")
WINNER_RE = re.compile(r"^W(\d+)$")
LOSER_RE = re.compile(r"^RU(\d+)$")
LOSER_ALT_RE = re.compile(r"^L(\d+)$")


def _knockout_outcome_label(match_num: int, outcome: str) -> str:
    role = "Winner" if outcome == "winner" else "Loser"
    if 101 <= match_num <= 102:
        return f"{role} of semi-final {match_num - 100}"
    if 97 <= match_num <= 100:
        return f"{role} of quarter-final {match_num - 96}"
    if 89 <= match_num <= 96:
        return f"{role} of Round of 16 (match {match_num})"
    if 73 <= match_num <= 88:
        return f"{role} of Round of 32 (match {match_num})"
    return f"{role} of match {match_num}"


def placeholder_display_name(code: str | None) -> str:
    if not code:
        return "To be decided"

    text = code.strip()
    if not text:
        return "To be decided"

    group_pos = GROUP_POSITION_RE.fullmatch(text)
    if group_pos:
        pos, group = group_pos.groups()
        return f"{ORDINALS[int(pos)]} in Group {group}"

    best_third = BEST_THIRD_RE.fullmatch(text)
    if best_third:
        letters = ", ".join(best_third.group(1))
        return f"Best 3rd-place team (groups {letters})"

    alt_group = ALT_GROUP_RE.fullmatch(text)
    if alt_group:
        group, pos = alt_group.groups()
        return f"{ORDINALS[int(pos)]} in Group {group}"

    winner = WINNER_RE.fullmatch(text)
    if winner:
        return _knockout_outcome_label(int(winner.group(1)), "winner")

    loser = LOSER_RE.fullmatch(text) or LOSER_ALT_RE.fullmatch(text)
    if loser:
        return _knockout_outcome_label(int(loser.group(1)), "loser")

    return text
